Lazy sublanguage

mattjj opened the pull request:
**TLDR**

Before this commit, this computation would avoid materializing the iota (`np.arange`) array at trace time:

```python
@jit
def f(x):
  m, n = x.shape
  return x + np.arange(n)
```

But this one would materialize the iota array at trace time and stage it into the computation as a potentially large array constant:

```python
@jit
def f(x):
  m, n = x.shape
  return x + np.arange(m)[:, None]
```
The difference is that previously operations like broadcasts, transposes, and singleton-dimension-adding reshapes (as above) would force otherwise lazy values to be materialized, while after this commit broadcasts, transposes, and those reshapes are all lazy operations that only update metadata and reuse the same device buffer as their input rather than compiling and executing XLA computations and producing new buffers.

Also, `np.eye` and `np.tri` become lazy (in addition to `np.zeros`, `np.ones`, `np.full`).

Finally, this PR replaces the ad-hoc "lazy device constant" system.

In an earlier version of this PR, I had also included the feature that lazy expressions would be fused into eager mode op-by-op compiled computations. After thinking it through with @hawkinsp we decided to split that functionality out into a follow-up PR, if we decide to include it at all.

**TODO**
- [x] add a way to make identities and triangular matrices
- [x] systematic tests
- [x] make `jit` force its arguments (see [comment below](https://github.com/google/jax/pull/1668#issuecomment-561699616))
- [x] microbenchmarks
- [x] write implementation notes below

**Follow-up work** for subsequent PRs
- [x] check that this resolves the issues underlying #1640 (or tweak it to resolve them) [EDIT: actually it's already fixed by #1848]
- [ ] fuse lazy expressions into eager op-by-op computations
- [ ] consider adding dtype conversion to the lazy language
- [ ] remove use of `lax.tie_in` in as many places as possible
- [ ] leverage laziness in `pmap` dispatch
- [ ] also handle reshapes that remove singleton dimensions

## Design idea

The basic idea is to introduce a kind of lazy sublanguage. Each DeviceArray carries an expression in this lazy sublanguage, and when we apply certain operations we'll produce a result that refers to the same underlying buffer as the input but has an updated expression (rather than compiling and executing an XLA computation to produce a new result buffer). Only when we apply functions to the array that don't exist in the lazy sublanguage will we compile and execute an XLA computation (taking into account the any lazy expressions on the inputs).

We want a proper _sub_-language rather than, say, modeling the full jaxpr language, because that way we sidestep a tough tradeoff: if the lazy sublanguage can only express "cheap" operations, meaning operations we're happy to stage separately into multiple downstream consumers (and hence could be evaluated multiple times), then we don't need to worry about work sharing. If instead we had expensive computations, which we might not want to evaluate more than once, then we would need a more complex system that attempted to share work between evaluations of lazy expressions. (It gets complex because when we evaluate a lazy subexpression we'd need to decide which of its intermediates to materialize, and then update all equivalent lazy subexpressions in the system with the materialized values we computed.) Unfortunately "cheap" is hard to define precisely without understanding XLA better (on all backends), but the kinds of operations outlined above are ones that we expect can be fused into consumers, and so are "cheap" because the values of these expressions may never be materialized at all.

Because we're attaching expressions to DeviceArrays, this design operates "underneath" all the tracing logic: it sits under the impl rules. In other words, as the toy model below makes clear, it's something one can do in any numerical library, and all of JAX's tracing and transformation machinery sits on top unmodified.

## Lazy sublanguage

Other than being able to express the broadcasts, reshapes, and transposes we want, some design criteria for the language are:
1. cheap expressions (e.g. expressions XLA fuses into consumers, avoiding direct evaluation)
2. fixed-size canonical representation (i.e. it doesn't grow as we apply more broadcast/reshape/transpose operations) so to get more cache hits and for fast expression manipulation

Here's the abstract syntax in terms of AST constructors:

```haskell
data LazyExpr = LazyExpr Input Shape Dims
data Input = ArrayVar
           | Iota Dtype Size
           | Eye Dtype Shape Offset
           | Tri Dtype Shape Offset
           | Delta Shape Axes

type Shape = [Int]
type Dims = [Maybe Int]
type Size = Int
type Offset = Int
type Axes = [Int]
```

There are two components to a LazyExpr: an input and a reindexing specification. The input represents a base array to which the reindexing specification is applied.

An input can represent an array constructor (`Iota`, `Eye`, etc) or it can be an `ArrayVar` which encodes that the base array is some exogenous array value. (These LazyExprs are attached to DeviceArrays, so when the input part of the expression is `ArrayVar` that basically means the associated device buffer is the input, while if the input is an array constructor than the associated `device_buffer` field of the DeviceArray should be set to the sentinel value `xla.device_constant`.)

The reindexing specification encodes the shape of the final result and a list of dimensions, which are integers or Nones. The integer entries take on values 0, 1, ..., N-1 where N is the rank of the input array, and encode where the axes of the input array are to be mapped in the final output. When an entry is None that indicates that the corresponding axis of the result is a broadcasted one.

The corresponding AST constructors in Python look like

```python
LazyExpr = namedtuple('LazyExpr', ['input', 'shape', 'dims'])
LazyArrayVar = namedtuple('ArrayVar', [])
LazyIota = namedtuple('Iota', ['dtype', 'size'])
LazyEye = namedtuple('Eye', ['dtype', 'shape', 'offset'])
LazyTri = namedtuple('Tri', ['dtype', 'shape', 'offset'])
LazyDelta = namedtuple('Delta', ['dtype', 'shape', 'axes'])
```

Here are some examples of lazy expressions and the arrays they represent:

```python
LazyExpr(input=Iota(dtype=dtype('float32'), size=3), shape=(3, 4), dims=(0, None))
DeviceArray([[0., 0., 0., 0.],
             [1., 1., 1., 1.],
             [2., 2., 2., 2.]], dtype=float32)

LazyExpr(input=Iota(dtype=dtype('float32'), size=3), shape=(4, 3), dims=(None, 0))
DeviceArray([[0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.],
             [0., 1., 2.]], dtype=float32)
```

See xla.py for a numpy-based interpreter of this language.

## Toy model

Here's a model showing the main idea (except for the changes to op-by-op and jit/pmap logic in xla.py).

```python
from collections import namedtuple
from functools import partial
import itertools as it

import numpy as onp
from jax.util import unzip2

LazyExpr = namedtuple('LazyExpr', ['input', 'shape', 'dims'])
ArrayVar = namedtuple('ArrayVar', [])
Iota = namedtuple('Iota', ['dtype', 'size'])


class Array(object):
  __slots__ = ['buf', 'lexpr']
  def __init__(self, buf, lexpr):
    self.buf = buf
    self.lexpr = lexpr

def materialize(arr):
  input_, shape, dims = arr.lexpr

  t = type(input_)
  if t is ArrayVar:
    x = arr.buf
  elif t is Iota:
    assert arr.buf is None
    x = onp.arange(input_.size, dtype=input_.dtype)
  else:
    assert False

  bcast_dims, perm = unzip2((i, d) for i, d in enumerate(dims) if d is not None)
  transposed = onp.transpose(x, perm)
  out = _broadcast_in_dim(transposed, shape, bcast_dims)

  return out

# NumPy implementation of lax.broadcast_in_dim (from lax_reference.py)
def _broadcast_in_dim(operand, shape, bcast_dims):
  inshape = tuple(1 if i not in bcast_dims else d for i, d in enumerate(shape))
  return onp.broadcast_to(onp.reshape(operand, inshape), shape)


# making an array (with a trivial identity reindex expression)
def make_array(buf):
  lazy_expr = LazyExpr(ArrayVar(), buf.shape, tuple(range(buf.ndim)))
  return Array(buf, lazy_expr)

# making an iota
def make_iota(dtype, size):
  lazy_expr = LazyExpr(Iota(dtype, size), (size,), (0,))
  return Array(None, lazy_expr)

# now for operations! first, a lazy broadcasting operation, modeling XLA's.
def lazy_broadcast(arr, shape, bcast_dims):
  lazy_expr = arr.lexpr
  new_dims = [None] * len(shape)
  for i, d in enumerate(bcast_dims):
    new_dims[d] = lazy_expr.dims[i]
  new_lazy_expr = LazyExpr(lazy_expr.input, shape, new_dims)
  return Array(arr.buf, new_lazy_expr)

# second, a transpose (HLO transpose uses "where it comes from" encoding)
def lazy_transpose(arr, perm):
  lazy_expr = arr.lexpr
  new_shape = [lazy_expr.shape[i] for i in perm]
  new_dims = [lazy_expr.dims[i] for i in perm]
  new_lazy_expr = LazyExpr(lazy_expr.input, new_shape, new_dims)
  return Array(arr.buf, new_lazy_expr)
```

## Micro-benchmarks

This PR adds some work to the op-by-op dispatch path, so we want to check that there isn't a significant performance regression. In the absence of proper performance regression tests, I just did some quick checks by hand.

On master (CPU):
```
In [1]: from jax import lax
In [2]: timeit -r 100 -n 1000 lax.add(1, 1).block_until_ready()
92.9 µs ± 8.3 µs per loop (mean ± std. dev. of 100 runs, 1000 loops each)
In [3]: x = lax.add(1, 1)
In [4]: timeit -r 100 -n 1000 lax.add(x, x).block_until_ready()
46.4 µs ± 5.11 µs per loop (mean ± std. dev. of 100 runs, 1000 loops each)
```

On branch (CPU):
```
In [1]: from jax import lax
In [2]: timeit -r 100 -n 1000 lax.add(1, 1).block_until_ready()
94.2 µs ± 9.03 µs per loop (mean ± std. dev. of 100 runs, 1000 loops each)
In [3]: x = lax.add(1, 1)
In [4]: timeit -r 100 -n 1000 lax.add(x, x).block_until_ready()
47.1 µs ± 5.36 µs per loop (mean ± std. dev. of 100 runs, 1000 loops each)
```

I did something similar with `jit(lax.add)` and there wasn't any movement there either.

## Implementation  notes

- The lazy language itself (namely the AST constructor namedtuples, a NumPy-based evaluator `eval_lexpr`, and an XLA-based interpreter `stage_lexpr`) is added in lazy.py.
- The impl rules of relevant primitives are updated in lax.py, so that e.g. a broadcast on a DeviceArray only updates metadata (as in the toy model above).
- The DeviceArray class in xla.py is updated to keep a `_lazy_expr` attribute.
- I made a couple explicit sentinels, like `device_constant`, rather than using `None` in some places.
- DeviceConstant, FIlledConstant, IotaConstant, EyeConstant are all removed.

Fixes #1909, and incidentally fixes #1431 because I happened to be updating that code.

mattjj commented:
Running the mnist_vae.py example, I noticed something unfortunate: we were compiling the update loop (`run_epoch`) twice!

First call's signature:
```
WARNING:absl:Compiling run_epoch for args (ArgSpec(aval=ShapedArray(uint32[2]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(2,), dims=(0,)), xla_shape=u32[2]{0}), ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(0, 1)), xla_shape=f32[784,512]{1,0}), ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[10,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10, 512), dims=(0, 1)), xla_shape=f32[10,512]{1,0}), ArgSpec(aval=ShapedArray(float32[10,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10, 512), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(None,)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[512,784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 784), dims=(0, 1)), xla_shape=f32[512,784]{1,0}), ArgSpec(aval=ShapedArray(float32[512,784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 784), dims=(None, None)), xla_shape=f32[]), ArgSpec(aval=ShapedArray(float32[784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784,), dims=(0,)), xla_shape=f32[784]{0}), ArgSpec(aval=ShapedArray(float32[784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784,), dims=(None,)), xla_shape=f32[])).
```

Second and subsequent calls' signatures:
```
WARNING:absl:Compiling run_epoch for args (ArgSpec(aval=ShapedArray(uint32[2]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(2,), dims=(0,)), xla_shape=u32[2]{0}), ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(0, 1)), xla_shape=f32[784,512]{1,0}), ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(0, 1)), xla_shape=f32[784,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[512,10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 10), dims=(0, 1)), xla_shape=f32[512,10]{1,0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[10]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10,), dims=(0,)), xla_shape=f32[10]{0}), ArgSpec(aval=ShapedArray(float32[10,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10, 512), dims=(0, 1)), xla_shape=f32[10,512]{1,0}), ArgSpec(aval=ShapedArray(float32[10,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(10, 512), dims=(0, 1)), xla_shape=f32[10,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 512), dims=(0, 1)), xla_shape=f32[512,512]{1,0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512,), dims=(0,)), xla_shape=f32[512]{0}), ArgSpec(aval=ShapedArray(float32[512,784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 784), dims=(0, 1)), xla_shape=f32[512,784]{1,0}), ArgSpec(aval=ShapedArray(float32[512,784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(512, 784), dims=(0, 1)), xla_shape=f32[512,784]{1,0}), ArgSpec(aval=ShapedArray(float32[784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784,), dims=(0,)), xla_shape=f32[784]{0}), ArgSpec(aval=ShapedArray(float32[784]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784,), dims=(0,)), xla_shape=f32[784]{0})).
```

The difference is in the broadcasts, like this pair of entries:
```
ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(None, None)), xla_shape=f32[]),
```

```
ArgSpec(aval=ShapedArray(float32[784,512]), lazy_expr=LazyExpr(input=ArrayVar(), shape=(784, 512), dims=(0, 1)), xla_shape=f32[784,512]{1,0}),
```

Because of [the `np.zeros` used to initialize optimizers in optimizers.py](https://github.com/google/jax/blob/17813eab20604f89cccdd45ba06431b276d73839/jax/experimental/optimizers.py#L216), on the first call to `run_epoch` we're fusing the formation of those lazy zeros into the `run_epoch` computation, meaning we compile for a signature with scalar arguments and broadcasts on the inputs. Then on the second call to `run_epoch` we need to recompile for dense array inputs. (If we add an `xla.force` call to the zeros creation, like `xla.force(np.zeros_like(x0))` to the `init` function of `momentum` in optimizers.py, we don't have to recompile, as on master.)

Two possible solutions:
1. make `jit` force its arguments, as on master (while still keeping op-by-op application and `jit` closure nonstrict)
2. change optimizers.py, and similar user code, to call `xla.force` in cases like this

Discussing with @hawkinsp, we think Option 1 sounds like a better heuristic. (As an extension,  if we want it, we could add a `nonstrict_argnums` to `jit` to control this behavior, maybe useful for library writers.) Forcing users to think about laziness and change their code, as we'd need to change optimizers.py, is costly, and we don't yet have real use cases where that this strict-jit policy would be problematic.

hawkinsp commented on jax/lax/lax.py:
```diff
@@ -1055,46 +1057,51 @@ def full(shape, fill_value, dtype=None):
-def broadcasted_eye(dtype, shape, axes):
-  if not isinstance(axes, (list, tuple)) or not len(axes) >= 2:
-    raise TypeError("make_diagonal `axes` must be a tuple with len at least 2.")
+def eye(dtype, shape, offset):
```
These need comments (although I admit `eye` was lacking one previously.)

Are they intended as public `lax` APIs? I would lean towards: "no" given we may be able to remove their special nature sometime soon.

mattjj replied:
I agree not to make them public.

mattjj replied:
Done.

mattjj replied:
Should I add underscores?

hawkinsp replied:
It's probably fine as is, although it might be worth noting this in the docstrings.

hawkinsp commented on jax/lax/lax.py:
```diff
@@ -2474,6 +2496,24 @@ def _reshape_impl(operand, new_sizes, dimensions, old_sizes):
     return xla.apply_primitive(reshape_p, operand, new_sizes=new_sizes,
                                dimensions=dimensions, old_sizes=old_sizes)
 
+def _is_singleton_reshape(old, new):
```
Comment. What is a singleton reshape?

mattjj replied:
Done.

hawkinsp commented on jax/lax/lax.py:
```diff
@@ -4024,100 +4073,6 @@ def _tie_in_batch_rule(batched_args, batch_dims):
 masking.masking_rules[tie_in_p] = lambda vals, logical_shapes: vals[1]
 
 
-### constants
```
Nice!

mattjj replied:
Woo!

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -1744,19 +1744,11 @@ def eye(N, M=None, k=None, dtype=None):
   lax._check_user_dtype_supported(dtype, "eye")
   dtype = float_ if dtype is None else dtype
   M = N if M is None else M
+  k = int(k or 0)
```
Put the `k=0` on the argument?

mattjj replied:
Done.

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -1768,16 +1760,23 @@ def identity(n, dtype=None):
-    *(x for x in (start, stop, step) if x is not None))
-  return onp.arange(start, stop=stop, step=step, dtype=dtype)
+    return lax.iota(dtype, start)  # avoids materializing
+  elif step is None or onp.issubdtype(_dtype(step), onp.integer):
```
Use `issubdtype`? I removed all uses of `onp.issubdtype` from JAX, so I would prefer we don't reintroduce them. (They are problematic because they no longer do the same thing in some cases, although I admit this is not one of them.)

mattjj replied:
Amazing catch, thank you! Done.

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -1768,16 +1760,23 @@ def identity(n, dtype=None):
+    step = 1 if step is None else int(step)
+    quot, rem = divmod(stop - start, step)
+    num = quot + bool(rem)
+    return step * lax.iota(dtype, num) + start
```
Just to confirm my understanding: this always materializes the `iota`, correct?

mattjj replied:
Yes

hawkinsp commented on jax/numpy/lax_numpy.py:
```diff
@@ -2761,7 +2755,7 @@ def _index_to_gather(x_shape, idx):
   start_index_map = []
 
-  gather_indices = zeros((0,), dtype=int32)
+  gather_indices = onp.zeros((0,), dtype=int32)
```
Why this change? Revert?

mattjj replied:
It was causing an extra pointless compile. I'll check if it still is.

mattjj replied:
Yeah, it can cause us to compile and execute an extra `lax.broadcast_in_dim` versus just doing a device transfer the first time this value is consumed (and we still have to do the same sized device transfer when we execute the broadcast_in_dim). There's no downside to using `onp.zeros` here.

hawkinsp replied:
It's worth a comment then so no-one accidentally reverts it.

(Although really this whole function should be under `jit` and this should be staged into that `jit` computation.)

mattjj replied:
WDYT?

mattjj replied:
Oops, my browser was out of sync and so my "WDYT?" was mistimed.

Will comment!

mattjj replied:
Done.

Agree that this whole function, as with others, should be under `jit`, though I'm hesitant to do that without even some rudimentary performance regression tests.

hawkinsp commented on jax/tools/jax_to_hlo.py:
```diff
@@ -122,7 +122,7 @@ def jax_to_hlo(fn, input_shapes, constants=None):
                        'the default layout is allowed.' % str(shape))
 
-    args.append(np.zeros(shape.dimensions(), dtype=shape.numpy_dtype()))
+    args.append(onp.zeros(shape.dimensions(), dtype=shape.numpy_dtype()))
```
Why these changes?

mattjj replied:
Reverted.

hawkinsp commented on jax/util.py:
```diff
@@ -219,3 +219,9 @@ def get_module_functions(module):
       module_fns.add(attr)
   return module_fns
+
+def subvals(lst, replace):
```
Please document.

(I know this moved from `lax.py`, but I didn't know what it did there either and I'm still hazy about what the name means.)

mattjj replied:
Done.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
+      xla_shape = None
+    else:
+      xla_shape = x.device_buffer.shape()
+    return aval, x._lazy_expr, xla_shape
```
It did originally, but that caused measurable overheads compared to using raw tuples. I'll add a comment.

mattjj replied:
Actually, it is documented a couple lines up. I added another sentence to be extra clear.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -633,7 +805,13 @@ def __init__(self, aval, device_buffer):
+        self._npy_value = eval_lazy_expr(self._lazy_expr, None)
+      else:
+        temp = force(self)
+        self.device_buffer = temp.device_buffer
```
Good thinking, I'll change the policy not to keep the device-side value.

mattjj replied:
Done.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -138,31 +281,45 @@ def abstractify(x):
-  compiled = built_c.Compile(compile_options=xb.get_compile_options(),
-                             backend=xb.get_backend(backend))
+  built = _primitive_computation(prim, *arg_specs, **params)
+  logging.debug(built.GetHloText())
```
Wow good catches, I'm a logging newbie. I'll kill this.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -819,26 +1010,16 @@ def _foil_cse(c, x):
-    opts = xb.get_compile_options(device_assignment=device_assignment)
-    compiled = c.Build(xla_const).Compile((), opts, backend=xb.get_backend(backend))
-    return compiled.Execute(())
+def force(x):
```
Done, minimalistically. It's not exposed in api.py right now; it's only used internally. Still I can imagine advertising it more in the future.

mattjj replied:
WDYT? Needs more docstring?

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
+  return x
+
+
+ArgSpec = namedtuple("ArgSpec", ["aval", "lazy_expr", "xla_shape"])
```
Done.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
+  new_dims = tuple(lazy_expr.dims[i] for i in perm)
+  return LazyExpr(lazy_expr.input, new_shape, new_dims)
+
+def eval_lazy_expr(lazy_expr, x):
```
Yes: `x` basically models the environment, which (due to the sublanguage's simplicity) is either empty or has a single array in it. A reason to keep it separate is that (in the case of `stage_lazy_expr`) the value is not available when the expression is constructed (and `eval_lazy_expr` here closely follows `stage_lazy_expr`).

mattjj replied:
I added docstrings. WDYT?

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
 
 
+### lazy sub-language
+
```
Done.

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
+LazyArrayVar = namedtuple('ArrayVar', [])
+LazyIota = namedtuple('Iota', ['dtype', 'size'])
+LazyEye = namedtuple('Eye', ['dtype', 'shape', 'offset'])
+LazyTri = namedtuple('Tri', ['dtype', 'shape', 'offset'])
```
Eye and Tri model exactly np.eye and np.tri. Delta models Kronecker delta arrays.

I think this is a borderline case for documentation: it's easy for an interested party to figure out exactly what these things do by looking at the code. I'll add very short descriptions.

I'm not going to add a TODO because we could apply a "TODO: always be improving" everywhere in the code. I would add a TODO if we had a specific plan to make these better, but we don't. None of the more generic mechanisms I've prototyped or thought about would be a good fit here (because they'd be heavyweight, or would require making jaxprs hashable, etc).

mattjj replied:
Done.

mattjj replied:
I thought about it more, and what I said was wrong. This is a good case for documentation. Sorry.

I still think the same about the TODO though :)

mattjj replied:
I added I added this sentence to the main comment in this section:

> For the array constructor expressions, Iota builds a 1D sequence [0, 1, ..., N-1], Eye builds a 2D array with ones on a (possibly offset) diagonal and zeros elsewhere (like numpy.eye), Tri builds a triangular matrix with ones on and below a diagonal and zeros elsewhere (like numpy.tri), and Delta builds a Kronecker delta array with ones along its multidimensional main diagonal and zeros elsewhere.

I also added same-line comments on the namedtuple definition lines.

WDYT?

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -56,6 +56,148 @@ def _map(f, *xs): return tuple(map(f, *xs))
 def identity(x): return x
 
 
+### lazy sub-language
```
Hmm what module do you suggest? `jax.util.lazy`? It's a bit xla.py-specific because of `stage_lazy_expr` using the XLA computation builder, and it's only consumed here.

mattjj replied:
I moved it to its own module!

mattjj commented on jax/core.py:
```diff
@@ -116,14 +116,18 @@ def __init__(self, val):
+    if self.hash is None:
+      return self.val is other.val
+    else:
+      return (self.val == other.val or
```
I'll revert this change for now.

mattjj replied:
Done.

mattjj commented on jax/interpreters/pxla.py:
```diff
@@ -414,9 +417,10 @@ def xla_pmap_impl(fun, *args, **params):
 
-  abstract_args = map(xla.abstractify, args)
+  # TODO(mattjj): support laziness here, don't just force every argument
+  avals = [xla.abstractify(xla.force(x)) for x in args]
```
This is me being dumb! Thank you for the catch.

mattjj replied:
This was just a mistake, but it left me wondering: where are we `force`ing arguments to `pmap` computations? It turns out that they're getting forced in `_shard_device_array`, which calls `_unstack`.

mattjj commented:
I really appreciate the thorough review, @hawkinsp. Not only did you catch several mistakes and make several improvements, but also your documentation suggestions are good and I'm just a curmudgeon about it. This PR is getting much better already :)

mattjj commented:
Thanks! I made added underscores to `lax._eye`, `lax._delta`, and `lax._tri` and improved their docstrings.

mattjj committed:
implement lazy sublanguage

Before this commit, this computation would avoid materializing the iota
array at trace time:

  @jit
  def f(x):
    m, n = x.shape
    return x + np.arange(n)

But this one would materialize the iota array at trace time and stage it
into the computation as a potentially large array constant:

  @jit
  def f(x):
    m, n = x.shape
    return x + np.arange(m)[:, None]

The difference is that previously operations like broadcasts,
transposes, and reshapes that add singleton dimensions (as above) would
force otherwise lazy values to be materialized, while after this commit
broadcasts, transposes, and reshapes are all lazy operations that only
update metadata on their input rather than compiling and executing XLA
computations and producing new buffers.

Also, np.eye and np.tri become lazy (in addition to np.zeros, np.ones, np.full).

This commit replaces the ad-hoc "lazy device constant" system, which was
used to get the simpler behavior in the first example above.

Incidentally fixes #1431

See https://github.com/google/jax/pull/1668 for more.

gnecula commented:
It would be really nice to pull much of the PR descriptions into the documentation, otherwise that info may be hard to find.

mattjj commented:
You're right. The "Lazy Sublanguage" section [appears as a comment in the source code](https://github.com/google/jax/blob/e51b6b34e15fca8d108de10617c673321202a1fe/jax/lazy.py#L47) but I think you probably meant the jax.readthedocs.io documentation. I'll add it to my todos (documentation on this as well as on the device placement policy stuff we discussed yesterday).
